File: week-1/day5-handson/js/src/userRepository.js
/**
 * userRepository.js
 *
 * Defines the INTERFACE (contract) that any user repository
 * must satisfy — real DB or fake in-memory version.
 *
 * In production you would inject a RealUserRepository
 * that talks to PostgreSQL / MongoDB / etc.
 * In tests you inject a FakeUserRepository that stores
 * users in a plain JavaScript Map — no DB required.
 *
 * This pattern is called "Dependency Injection" and it is
 * what makes the Fake pattern possible.
 */

/**
 * REAL repository — would normally call a database.
 * Not used in tests — only here to show the production shape.
 */
class RealUserRepository {
  async save(user) {
    // db.query('INSERT INTO users ...', user);
    throw new Error('RealUserRepository: not implemented in this exercise');
  }

  async findById(id) {
    // return db.query('SELECT * FROM users WHERE id = ?', [id]);
    throw new Error('RealUserRepository: not implemented in this exercise');
  }

  async findAll() {
    // return db.query('SELECT * FROM users');
    throw new Error('RealUserRepository: not implemented in this exercise');
  }

  async deleteById(id) {
    // db.query('DELETE FROM users WHERE id = ?', [id]);
    throw new Error('RealUserRepository: not implemented in this exercise');
  }
}

/**
 * FAKE repository — stores users in memory.
 *
 * This has REAL logic (save actually saves, find actually finds)
 * but zero infrastructure. It is faster and simpler than the
 * real thing, which makes it perfect for unit tests that need
 * to verify multiple operations in sequence.
 *
 * A Fake is different from a Stub (which only hard-codes return
 * values) because it carries working business logic.
 */
class FakeUserRepository {
  constructor() {
    // Simple in-memory store: id → user object
    this.store = new Map();
  }

  async save(user) {
    if (!user || !user.id) throw new Error('User must have an id');
    this.store.set(user.id, { ...user }); // store a copy
  }

  async findById(id) {
    return this.store.get(id) ?? null;
  }

  async findAll() {
    return Array.from(this.store.values());
  }

  async deleteById(id) {
    const existed = this.store.has(id);
    this.store.delete(id);
    return existed; // true if something was actually deleted
  }
}

module.exports = { RealUserRepository, FakeUserRepository };

File: week-1/day5-handson/python/src/user_repository.py
# user_repository.py
#
# Defines the contract (interface) that any user repository
# must satisfy — real DB implementation or Fake in-memory version.
#
# In production you inject a RealUserRepository that talks
# to a real database. In tests you inject a FakeUserRepository
# that stores users in a plain Python dict — no DB required.
#
# This is DEPENDENCY INJECTION: the service doesn't care which
# repository it receives, as long as it satisfies the interface.


class RealUserRepository:
    """
    Production repository — would normally call a real database.
    Not used in tests — only here to show the production shape.
    """

    def save(self, user: dict) -> None:
        # db.execute("INSERT INTO users ...", user)
        raise NotImplementedError("RealUserRepository: not implemented in this exercise")

    def find_by_id(self, user_id: str) -> dict | None:
        # return db.execute("SELECT * FROM users WHERE id = ?", [user_id])
        raise NotImplementedError("RealUserRepository: not implemented in this exercise")

    def find_all(self) -> list[dict]:
        # return db.execute("SELECT * FROM users")
        raise NotImplementedError("RealUserRepository: not implemented in this exercise")

    def delete_by_id(self, user_id: str) -> bool:
        # db.execute("DELETE FROM users WHERE id = ?", [user_id])
        raise NotImplementedError("RealUserRepository: not implemented in this exercise")


class FakeUserRepository:
    """
    FAKE repository — stores users in a plain Python dict.

    This has REAL logic (save actually saves, find actually finds)
    but uses in-memory storage instead of a database. It is faster
    and simpler than the real thing, making it perfect for unit
    tests that need to verify multiple operations in sequence.

    A Fake differs from a Stub:
      - Stub: always returns the same hard-coded value.
      - Fake: has working logic — what you save, you can retrieve.
    """

    def __init__(self):
        # Simple in-memory store: id → user dict
        self._store: dict[str, dict] = {}

    def save(self, user: dict) -> None:
        if not user or 'id' not in user:
            raise ValueError("User must have an 'id' field")
        # Store a copy so external mutations don't affect stored data
        self._store[user['id']] = dict(user)

    def find_by_id(self, user_id: str) -> dict | None:
        return self._store.get(user_id)

    def find_all(self) -> list[dict]:
        return list(self._store.values())

    def delete_by_id(self, user_id: str) -> bool:
        if user_id in self._store:
            del self._store[user_id]
            return True
        return False  # nothing was deleted
